### PR TITLE
Save all metrics for in-depth analysis

### PR DESCRIPTION
**Pull request by developer0hye**
### Summary
📊🚀🔧

Modified `results_dict` function in `ultralytics/utils/metrics.py` to include class-specific metrics for detection. This improves the logging and visualization of the model performance.

> _`results_dict` grows_
> _Class metrics added for each_
> _Autumn harvest of data_

### Walkthrough
*  Add class-specific metrics to results dictionary ([link](https://github.com/ultralytics/ultralytics/pull/5673/files?diff=unified&w=0#diff-fb567871ecb755681f83f6e1edeb903d3a5757316d9caa1a28bd2be993bf4fc8L758-R764))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 📊 Key Changes
- Extended the `results_dict` function in `utils/metrics.py` to include per-class metrics in the results dictionary.
- Adjusted the plotting index in `utils/plotting.py` to account for new metrics.

### 🎯 Purpose & Impact
The changes allow for a more granular analysis of model performance by providing individual class metrics, enabling users to pinpoint strengths and weaknesses of the machine learning model on a per-class basis 🎓. For users, this means more detailed insights into their model's accuracy and the ability to fine-tune it with greater precision 🔍.

### 🌟 Summary
This PR enhances performance analysis by saving all metrics per class for in-depth evaluation 📈, offering users a deeper understanding of their model's performance on individual classes.

### CODE REVIEW

**Comment by developer0hye** [comment]
image

image

image

**Comment by developer0hye** [comment]
Averaging metrics provides a simple overview of our model's performance, but this approach may omit details that are critical for in-depth analysis. I propose that we save both the class-specific metrics and their mean values, or at least offer this as an optional feature with `args`.

**Comment by glenn-jocher** [comment]
@developer0hye thank you for this valuable suggestion. Providing class-specific metrics is indeed a significant feature for understanding the nuances in model performance across different classes. We are committed to providing detailed metrics to facilitate the evaluation of the models' performance.

Having an option to save class-specific metrics and mean values can be beneficial for in-depth analysis. This could be done as you suggested with an optional feature controlled by a new argument that could be set when running your training script.

This way, if a user wants a more general view of the model performance, they can stick with the averaged metrics for a broad overview. And if they want a more detailed view for particular classes or for conducting more in-depth analysis, they can simply choose to save the class-specific metrics.

We appreciate your suggestion and it is something we will definitely take into consideration for future updates to our YOLOv8 repository. Please stay tuned for further improvements and additions. Your input aids in improving and refining our work, contributing to the YOLO community's overall advancement. Thank you!

**Comment by developer0hye** [comment]
image

This code has a bug related to plotting mean metrics on results.png.

I am working on it.

**Comment by developer0hye** [comment]
I solved the above issue.

image

**Comment by glenn-jocher** [comment]
@developer0hye thank you for your contribution to the YOLOv8 repository. From your report, it seems that the bug related to plotting mean metrics for validation loss on `results.png` has been resolved, which is excellent news!

Your efforts towards detecting and resolving this issue are greatly appreciated. It is contributions like these which help improve and fine-tune our models, thus directly benefiting the YOLO community as a whole. Thanks again for your dedication and hard work. 

Moving forward, anybody who was previously encountering this bug should no longer experience this issue thanks to your fix. We look forward to any further contributions you may make, and we encourage other members of the community to get involved in the development and improvement of YOLOv8 as well.

**Comment by developer0hye** [comment]
@glenn-jocher What about this PR?

**Comment by glenn-jocher** [comment]
@developer0hye thank you for your contribution. We are currently reviewing your pull request and its potential impact on the YOLOv8 project. We aim to provide constructive feedback and make decisions that best support the repository's evolution. We appreciate your patience in this process and laud your effort to improve the project. Please stay tuned for further updates on your PR.

**Comment by developer0hye** [comment]
@glenn-jocher Thanks for your team's sincere review about my PR!

**Comment by glenn-jocher** [comment]
@developer0hye You're welcome! We truly appreciate the effort you've put into your pull request. It's contributions like these that help drive our project forward. We're thoroughly reviewing your proposed changes to ensure it aligns with our project's direction. As always, we value your contribution to the YOLOv8 project. Thank you for your patience and for being part of our developer community.

**Comment by bsljsljy** [comment]
How to use tensorboard to check bn_weight in yolov8.

**Comment by bsljsljy** [comment]
How to use tensorboard to check bn_weight in yolov8.

**Comment by bsljsljy** [comment]
How to use tensorboard to check bn_weight in yolov8.

**Comment by bsljsljy** [comment]
> 嘿！要使用 TensorBoard 检查 YOLOv8 中的（批量归一化权重），您可以添加几行代码来在训练期间记录这些权重。下面是一个简单示例：`bn_weight`
> 
> ```python
> from torch.utils.tensorboard import SummaryWriter
> 
> # Initialize the TensorBoard writer
> writer = SummaryWriter('runs/your_experiment_name')
> 
> # Assuming 'model' is your YOLOv8 model and 'epoch' is your current epoch number
> for name, param in model.named_parameters():
>     if "bn" in name and "weight" in name:
>         writer.add_histogram(name, param, epoch)
> 
> writer.close()
> ```
> 
> 将此代码片段添加到您的训练循环中，其中您的 YOLOv8 模型。这会将批量归一化权重的直方图记录到 TensorBoard，您可以通过在终端中运行来查看。祝您编码愉快！😊`model``tensorboard --logdir=runs`

Thank you for your reply. I'm sorry that my programming skills are weak. Can you teach me in more detail the specific process of where the code should be added?

**Comment by glenn-jocher** [comment]
Hey there! No worries at all, we’re here to help. 😊 To add the code for tracking `bn_weight` with TensorBoard, you’ll usually do this inside your training loop, where your model is being trained over epochs. Here's a rough idea to guide you:

1. First, ensure you've initiated a `SummaryWriter` as shown in your snippet.
2. Then, find the part of your code that iterates through epochs - this is your training loop.
3. Inside this loop, after an epoch ends or even during it if you want more granular data, add the code snippet you shared. 

Here's a simplified example if your training loop looks something like this:

```python
for epoch in range(total_epochs):
    # Train your model...
    # model training code here

    # Now add the snippet for logging bn_weight
    for name, param in model.named_parameters():
        if "bn" in name and "weight" in name:
            writer.add_histogram(name, param, epoch)
            
    # Model validation, etc...

writer.close()  # Make sure to close the writer after the training loop
```

After this, once you run your training script, you can open TensorBoard to see the histograms of your batch normalization weights by running `tensorboard --logdir=runs` from the terminal or command prompt.

Hope this clears things up! Feel free to reach out if you have more questions. Happy coding! 🚀
